src/cached_object_store.rs

```diff
@@ -762,6 +817,286 @@ impl LocalCacheEntry for FsCacheEntry {
+                        "evictor: failed to get the metadata of the cache file: {}",
+                        err
+                    );
+                    continue;
+                }
+            };
+            let atime = metadata.accessed().unwrap_or(SystemTime::UNIX_EPOCH);
+            let path = entry.path().to_path_buf();
+            let bytes = metadata.len() as usize;
+
+            self.track_new_entry(path, bytes, atime, evict).await;
+        }
+    }
+
+    async fn track_new_entry(
+        &self,
+        path: std::path::PathBuf,
+        bytes: usize,
+        accessed_time: SystemTime,
+        evict: bool,
+    ) -> usize {
+        // record the new cache entry into the cache_entries, and increase the cache_size_bytes
+        self.cache_size_bytes
+            .fetch_add(bytes as u64, Ordering::SeqCst);
+        self.cache_entries
+            .lock()
+            .await
+            .insert(path.clone(), (accessed_time, bytes));
+
+        // record the metrics
+        self.db_stats
+            .object_store_cache_keys
+            .set(self.cache_entries.lock().await.len() as u64);
+        self.db_stats
+            .object_store_cache_bytes
+            .set(self.cache_size_bytes.load(Ordering::Relaxed));
+
+        if !evict {
+            return 0;
+        }
+
+        // if the cache size is still below the limit, do nothing
+        if self.cache_size_bytes.load(Ordering::Relaxed) <= self.max_cache_size_bytes as u64 {
+            return 0;
+        }
+        // TODO: check the disk space ratio here, if the disk space is low, also triggers evict.
+
+        // if the cache size exceeds the limit, evict the cache files in batch with the batch_factor,
+        // this may help to avoid the cases like triggering the evictor too frequently when the cache
+        // size is just slightly above the limit.
+        let mut total_bytes: usize = 0;
+        for _ in 0..self.batch_factor {
+            let evicted_bytes = self.maybe_evict_once().await;
+            if evicted_bytes == 0 {
+                return total_bytes;
+            }
+
+            total_bytes += evicted_bytes;
+        }
+
+        total_bytes
+    }
+
+    // find a file, and evict it from disk. return the bytes of the evicted file. if no file is evicted or
+    // any error occurs, return 0.
+    async fn maybe_evict_once(&self) -> usize {
+        let (target, target_bytes) = match self.pick_evict_target().await {
+            Some(target) => target,
+            None => return 0,
+        };
+
+        if let Err(err) = tokio::fs::remove_file(&target).await {
+            warn!("evictor: failed to remove the cache file: {}", err);
+            return 0;
+        }
+
+        debug!(
+            "evictor: evicted cache file: {:?}, bytes: {}",
+            target, target_bytes
+        );
+
+        // remove the entry from the cache_entries, and decrease the cache_size_bytes
+        self.cache_entries.lock().await.remove(&target);
```

Comment:
This seems vulnerable to (albeit pretty unlikely) race conditions where both the evictor and scan tasks choose to evict the same entry. For example, the evictor task we could re-insert an entry into `cache_entries` before the scanner hits this line, and then the entry would no longer be tracked. Maybe it's better to just hold a coarse-grained lock over `track_new_entry`. We'd be holding it across I/O operations, but that seems ok since it's just the scanner and evictor tasks that try to acquire the lock.

Reply:
this makes sense, i added a coarser lock in 5c343f0 to lock `cache_entries` and `cache_size_bytes` together.

this also fixed a bug: when a cached file is accessed another time, it should not increase `cache_size_bytes`, because the cache size is actually not changed.

Reply:
I'm having a hard time convincing myself that there aren't races with the current approach. For example, what prevents the following:

1. scanner task chooses to evict item E and deletes its file
2. cache creates a new cache file for item E and sends an accessed msg to tx
3. evictor task creates a new cache entry for E
4. scanner task deletes item E from cache_entries, even though its still cached

I was actually suggesting we just take the lock at the beginning of `track_new_entry` and hold it for the whole method. Alternatively, we could just have the scanner task add items to `tx` rather than try to track/evict items itself and then `track_entry_accessed` wouldn't have to do any synchronization. IMO this will make this code easier to reason about.

Reply:
oh, yes it's indeed an race issue 😲.

it's not considered an issue when the files are deleted on the disk but leaked in the memory map (it can be finally recycled on eviction). but it's bad when there're files on disk but not tracked in the in-memory map, it will never evict these files unless taking a intervally rescan approach like what juicefs did.

i kinda prefer using a coarsed lock when comparing with unify the logic behind `tx`, because the intention of introducing this `tx` is to avoid blocking the main task on eviction works, but while `rescan` is running, the `tx` is likely to be filled up in a big loop and likely to block the caller.

Reply:
> i kinda prefer using a coarsed lock when comparing with unify the logic behind tx, because the intention of introducing this tx is to avoid blocking the main task on eviction works, but while rescan is running, the tx is likely to be filled up in a big loop and likely to block the caller.

Fair enough. I suppose we could also set up separate queues for the main tasks and the scanner tasks to write to and `select!` from the evictor. Then the eviction metadata is stricty owned by the evictor task which is nice. But I'm ok with just using a coarse lock too.

Reply:
fixed in a4d1994

Reply:
One suggestion to simplify the code a bit - rather than adding a separate lock that we take from `track_entry_accessed`, we could instead crate an `Arc<Mutex<FsCacheEvictorInner>>` in `FsCacheEvictor::start`, which would require us to acquire its lock whenever we want to track another item. This way we don't have to make all the entries in `FsCacheEvictorInner` thread-safe which hurts readability a bit. You would also have to move `scan_entries` out of `FsCacheEvictorInner` but that seems ok.

Reply:
indeed, i like this approach of having a mental set to serialize the call of FsCacheEvictorInner

let me raise another pr on it.